File: subprojects/docs/src/samples/userguide/plugins/readOnlyManagedProperty/java/buildSrc/src/main/java/UrlProcess.java
import org.gradle.api.DefaultTask;
import org.gradle.api.provider.Property;
import org.gradle.api.tasks.TaskAction;

import java.net.URI;

public abstract class UrlProcess extends DefaultTask {
    // Use an abstract getter method
    abstract Property<URI> getUri();

    @TaskAction
    void run() {
        // Use the `uri` property
        System.out.println("Downloading " + getUri().get());
    }
}

File: subprojects/docs/src/samples/userguide/plugins/readOnlyManagedProperty/kotlin/buildSrc/src/main/kotlin/UrlProcess.kt
import org.gradle.api.DefaultTask
import org.gradle.api.tasks.TaskAction
import org.gradle.api.provider.Property
import java.net.URI

abstract class UrlProcess : DefaultTask() {
    // Use an abstract val
    abstract val uri: Property<URI>

    @TaskAction
    fun run() {
        // Use the `uri` property
        println("Downloading ${uri.get()}")
    }
}
